### PR TITLE
Migrate from Travis CI to GitHub Actions

### PR DESCRIPTION
**Pull request by suushiemaniac**
See title. I expect this PR to take some time based on trial-and-error checks, since it is arguably hard to debug CI run without actually running the entire CI.

The Ruby gems at `$HOME/.bundle` will be automatically cached by the `setup-ruby` action. The MySQL password of the Ubuntu installation also has to be manually adjusted (the MySQL server that comes with GitHub Actions has `root:root` instead of Travis'  `root:[empty]`) -> can we change our website Ruby code to use `root:root`?

Everything else should just be a 1:1 adaptation of the old Travis checks in the new GitHub Actions syntax.

### CODE REVIEW

**Comment by suushiemaniac** [comment]
Interesting observation: The CI currently fails because the `overcommit` checks demand that `git --config user.name` and `user.email` be set. In the GitHub Actions environment, they are not set by default.

**Comment by suushiemaniac** [comment]
Ready for review!

File: .travis-old.yml [hunk 1]
```diff
@@ -8,7 +8,7 @@ cache:
     - $HOME/.phantomjs
```
Comment (jonatanklosko) [inline .travis-old.yml:8]: Feel free to just remove this file altogether, we can always find it in the history :>

File: .github/workflows/ruby.yml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  push:
+    branches: [ master ]
+  pull_request:
+    branches: [ master ]
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    env:
+      DATABASE_URL: mysql2://root:@127.0.0.1:3306/wca_test
```
Comment (jonatanklosko) [inline .github/workflows/ruby.yml:20]: When `DATABASE_URL` env variable is specified it overrides any default configuration, so if you specify a URL with the default GH Actions password like `mysql2://root:root@127.0.0.1:3306/wca_test`, it should work without altering the password.

Also, I think GH Actions have mysql 5.7 by default, but ideally we would run mysql 8 to mirror the production server. I think one option would be to use the `services:` configuration, which allows for running arbitrary docker images there.

File: .github/workflows/ruby.yml [hunk 3]
```diff
@@ -0,0 +1,63 @@
+# This workflow uses actions that are not certified by GitHub.
+# They are provided by a third-party and are governed by
+# separate terms of service, privacy policy, and support
+# documentation.
+# This workflow will download a prebuilt Ruby version, install dependencies and run tests with Rake
+# For more information see: https://github.com/marketplace/actions/setup-ruby-jruby-and-truffleruby
```
Comment (jonatanklosko) [inline .github/workflows/ruby.yml:6]: Nit: this is pretty generic so I think there's not much value having it, when looking for details it's best to lookup the specific action at hand.

File: .github/workflows/ruby.yml [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+          bundler-cache: true # runs 'bundle install' and caches installed gems automatically
+      - uses: actions/setup-node@v2 # this also includes Yarn
+        with:
+          node-version: '12'
+      - uses: actions/cache@v2
+        with:
+          path: '**/node_modules'
+          key: ${{ runner.os }}-node_modules-${{ hashFiles('**/yarn.lock') }}
+          # restore-keys: |
+          #   ${{ runner.os }}-node_modules
```
Comment (jonatanklosko) [inline .github/workflows/ruby.yml:45]: Nit: leftover

**Review by jonatanklosko — COMMENTED** [review]
Beautiful! :tada: :heart: 

Just some minor comments, please try setting up mysql8 as per one of the comments and we should be ready to go!

File: .travis-old.yml (hunk 1 above)
Comment (suushiemaniac) [reply .travis-old.yml:8]: I was mostly keeping this as a reference for myself while working on this PR. Now that we have feature parity, I can indeed remove it :D

File: .github/workflows/ruby.yml (hunk 2 above)
Comment (suushiemaniac) [reply .github/workflows/ruby.yml:20]: According to https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md#mysql GitHub Actions runs mysql 8!

File: .github/workflows/ruby.yml (hunk 3 above)
Comment (suushiemaniac) [reply .github/workflows/ruby.yml:6]: This was actually part of the template provided by GitHub. Good catch, will remove!

File: .github/workflows/ruby.yml (hunk 2 above)
Comment (jonatanklosko) [reply .github/workflows/ruby.yml:20]: Oh, perfect!!!

File: .github/workflows/ruby.yml (hunk 4 above)
Comment (suushiemaniac) [reply .github/workflows/ruby.yml:45]: Not necessarily a leftover. `restore-keys` is pretty much a backup mechanism: If it doesn't find a cache stored under the key provided by `key`, it tries `restore-keys` in order as fallback. I left this in to debate whether we want to use fallbacks here, i.e. rely on older caches where the hash of `yarn.lock` doesn't match. Opinions?

Comment (jonatanklosko) [reply .github/workflows/ruby.yml:45]: Ah, yeah I think using an older cache should be fine, especially given the automated dependency updates we get.

**Comment by suushiemaniac** [comment]
Had some minor troubles with getting everything uploaded to Coveralls, (potential) fix is already uploaded and running through CI as I write this comment!

**Comment by suushiemaniac** [comment]
I have just realised that our previous Coveralls reports (via the Ruby `coveralls` gem) all cover the `spec` folder itself. Is this intended/desired?

SimpleCov itself filters the `spec` filter in [its base configuration](https://github.com/simplecov-ruby/simplecov/blob/main/lib/simplecov/profiles/test_frameworks.rb), implying that this shouldn't be tracked, I can fix our Simplecov to not track `/spec` but this would result in a significant coverage percentage drop (~96 to ~90). Please advise @jonatanklosko

**Comment by jonatanklosko** [comment]
> SimpleCov itself filters the spec filter in its base configuration, implying that this shouldn't be tracked. I can fix our Simplecov to not track /spec but this would result in a significant coverage percentage drop (~96 to ~90). Please advise @jonatanklosko

Ohh, interesting point! There are [different opinions](https://codecov.wordpress.com/2019/10/25/should-i-include-test-files-in-code-coverage-calculations/), but I think excluding the test files makes much sense, so I'm up for the new configuration.

I believe we are ready now, or do you want to investigate anything else?

**Comment by jonatanklosko** [comment]
Ah, feel free to enable the `restore-keys` as mentioned.

**Comment by suushiemaniac** [comment]
Oki, will do! I'll also try out the `rails` configuration of SimpleCov to see if that changes anything significant

**Comment by suushiemaniac** [comment]
Hm, using the `rails` configuration for SimpleCov really messes things up it seems. I will revert to the status quo *including the spec folder* and we can worry about handling of coverage in a separate PR to keep concerns clean and my mind sane.

**Comment by jonatanklosko** [comment]
Awesome work!

**Pull request merged**